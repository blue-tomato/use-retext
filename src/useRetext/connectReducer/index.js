import { mapValuesDeep } from '../../helpers';

export default (reducer, emitter, onCall) => {
  mapValuesDeep(reducer, ({ key, value, scope: reducerScope }) => {
    emitter.on(key, ({ scope: emitterScope, payload }) => {
      // Ensure that the reducer-scope starts at least with the emitter-scope
      if (!reducerScope.startsWith(emitterScope)) return;

      onCall({ key, scope: reducerScope, reducer: value, payload });
    });
  });
};
