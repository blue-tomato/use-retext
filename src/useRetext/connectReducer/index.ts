import { Emitter } from 'mitt';
import { mapValuesDeep } from '../../helpers';
import { Reducer, ReducerFuntion } from '../../Store';

interface OnCallObject {
  key: string;
  scope: string;
  reducer: ReducerFuntion;
  payload: any;
}

export default (reducer: Reducer, emitter: Emitter, onCall: (event: OnCallObject) => void) => {
  mapValuesDeep(reducer, ({ key, value, scope: reducerScope }) => {
    emitter.on(key, ({ scope: emitterScope, payload }) => {
      // Ensure that the reducer-scope starts at least with the emitter-scope
      if (!reducerScope.startsWith(emitterScope)) return;

      onCall({ key, scope: reducerScope, reducer: value, payload });
    });
  });
};
