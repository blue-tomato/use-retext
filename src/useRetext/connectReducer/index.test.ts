import mitt from 'mitt';
import connectReducer from '.';

const { store } = global;

describe('connectReducer', () => {
  it('connects reducer with emitter', () => {
    const emitter = mitt();

    connectReducer(store.reducer, emitter, (args) => {
      expect(args).toMatchSnapshot();
    });

    emitter.emit('increment', { scope: '' });
    emitter.emit('toggle', { scope: 'sideMenu' });
    emitter.emit('toggle', { scope: 'sideMenu.child' });
  });
});
