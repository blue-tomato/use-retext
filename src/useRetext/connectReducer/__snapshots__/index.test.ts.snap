// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectReducer connects reducer with emitter 1`] = `
Object {
  "key": "increment",
  "payload": undefined,
  "reducer": [Function],
  "scope": "",
}
`;

exports[`connectReducer connects reducer with emitter 2`] = `
Object {
  "key": "increment",
  "payload": undefined,
  "reducer": [Function],
  "scope": "sideMenu",
}
`;

exports[`connectReducer connects reducer with emitter 3`] = `
Object {
  "key": "toggle",
  "payload": undefined,
  "reducer": [Function],
  "scope": "sideMenu",
}
`;

exports[`connectReducer connects reducer with emitter 4`] = `
Object {
  "key": "toggle",
  "payload": undefined,
  "reducer": [Function],
  "scope": "sideMenu.child",
}
`;

exports[`connectReducer connects reducer with emitter 5`] = `
Object {
  "key": "toggle",
  "payload": undefined,
  "reducer": [Function],
  "scope": "sideMenu.child",
}
`;
